### PR TITLE
feat: KV Connect protocol spec

### PR DESCRIPTION
**Pull request by lucacasonato**
This commit adds a proper spec for the KV Connect protocol.

### CODE REVIEW

File: proto/kv-connect.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+metadata exchange and display an error message to the user.
+
+If client MUST verify that at least one of the entries in the `endpoints` array
+has a `consistency` property that is equal to `strong`. If no such entry is
+found, the client MUST fatally abort the metadata exchange and display an error
+message to the user.
+
+If the client is using protocol version 2 or higher, the client MUST parse each
+URL in the endpoints array and resolve it against the base URL of the metadata
+exchange request. If the client followed a redirect during the metadata exchange
```
Comment (igorzi) [inline proto/kv-connect.md:126]: For my education, what's changing in v2 that requires the URL parsing?

File: proto/kv-connect.md [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+- An _Atomic Write Request_ which is used to write some values to the database.
+
+#### Snapshot Read Request
+
+A _Snapshot Read Request_ is used to read some values from the database.
+
+<!-- A _Snapshot Read Request_ can either be "single shot" or "watching". A "single
+shot" _Snapshot Read Request_ is used to read a given set of values from the
+database once. A "watching" _Snapshot Read Request_ is used to read a given set
+of values from the database once and then receive a stream of  -->
```
Comment (igorzi) [inline proto/kv-connect.md:223]: There are a few commented out paragraphs. Did you mean to leave them in?

File: proto/kv-connect.md [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+The server MAY return a 5xx class HTTP status response at any time if the server
+is unable to process the request right now due to a temporary condition. The
+server SHOULD include a plain text body with a human readable error message in
+this case.
+
+If the access token is valid and a database is found, the server MUST respond
+with a 200 OK response. This response MUST include a `Content-Type` header with
+the value `application/json`. The text encoding of the response body MUST be
+UTF-8. The response body MUST be a JSON object that adheres to the the JSON
```
Comment (igorzi) [inline proto/kv-connect.md:95]: extra "the"

File: proto/kv-connect.md [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+
+<!-- If the `x-denokv-stream` header is included and has the value `true`, the server
+MUST perform the requested snapshot read operation. If the operation fails due
+to a malformed request, the server MUST respond with a 4xx class response
+containing a plain text body with a human readable error message. If the
+operation succeeds the server must respond with a 200 OK response. The response
+MUST include a `Content-Type` header with the value `...`. The response body
+MUST be a stream of Protobuf -->
+
+The server MUST perform the requested snapshot read operation. The the server
```
Comment (igorzi) [inline proto/kv-connect.md:269]: extra "the"

File: proto/kv-connect.md [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+adheres to the Protobuf schema for the `com.deno.kv.datapath.AtomicWriteRequest`
+message. If the request body is invalid, the server MUST respond with a 4xx
+class HTTP status response and a plain text body containing a human readable
+error message.
+
+The server MAY perform quota checks on the request. If the request is rejected
+due to quota limits, the server MUST respond with a 4xx class HTTP status
+response and a plain text body containing a human readable error message.
+
+The server MUST perform the requested atomic write operation. The the server
```
Comment (igorzi) [inline proto/kv-connect.md:352]: extra "the"

File: proto/kv-connect.md [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+response and a plain text body containing a human readable error message.
+
+The server MAY return a 5xx class HTTP status response at any time if the server
+is unable to process the request right now due to a temporary condition. The
+server SHOULD include a plain text body with a human readable error message in
+this case.
+
+The server MAY perform quota checks on the request. If the request is rejected
+due to quota limits, the server MUST respond with a 4xx class response and a
+plain text body containing a human readable error message.
```
Comment (igorzi) [inline proto/kv-connect.md:259]: How about rate limiting?

File: proto/kv-connect.md [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+has an eventually consistent view. If the request succeeds, the server MUST
+include the `ranges` field with a list of ranges that were read from the
+database. The order MUST be the same as the order of the ranges in the request.
+
+If the client fails to receive a response from the server due to a network
+error, or a 5xx class HTTP status, the client SHOULD retry the request using an
+exponential backoff strategy with unlimited retries.
+
+The client MUST verify that the response is a 200 OK response. If the response
+is a 4xx class error response, the client MUST fatally abort the request and
```
Comment (igorzi) [inline proto/kv-connect.md:286]: Similar comment to above about rate limiting. Should the server be able to respond with 429? and should the client interpret that as retry with backoff?

File: proto/kv-connect.md [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+client MUST verify that the response body adheres to the Protobuf schema for the
+`com.deno.kv.datapath.AtomicWriteResponse` message. If the response body is
+invalid, the client MUST fatally abort the request and display the error message
+to the user.
+
+If the response has a `status` field set to `AW_WRITE_DISABLED`, the client
+SHOULD perform a metadata exchange with the server to get a new list of
+endpoints, and then retry the request.
+
+If the response has a `status` field set to `AW_CHECK_FAILED`, the client MUST
```
Comment (igorzi) [inline proto/kv-connect.md:379]: I believe we also now return the conflict key for `AW_CHECK_FAILED`, which should also be mentioned here

File: proto/kv-connect.md (hunk 2 above)
Comment (lucacasonato) [reply proto/kv-connect.md:223]: Watch stuff is not quite ready yet - Heyang and me are still debating on the final API, so I left this out for the moment.

File: proto/kv-connect.md (hunk 1 above)
Comment (lucacasonato) [reply proto/kv-connect.md:126]: The `url` in the `endpoints` array can now be a URL relative to the metadata URL.

```jsonc
// Hosted at https://example.com/metadata/1234
{
  "version": 2,
  "endpoints": [
    {
      "url": "/data/v1", // resolved to https://example.com/data/v1
      "consistency": "strong"
    }
  ],
  // ...
}
```

File: proto/kv-connect.md (hunk 7 above)
Comment (lucacasonato) [reply proto/kv-connect.md:286]: I think the server should respond with a 503 then. It's easier if 4xx == never retry, and 5xx == retry with backoff.

File: proto/kv-connect.md (hunk 6 above)
Comment (lucacasonato) [reply proto/kv-connect.md:259]: I have added a note to the paragraph above about using 5xx responses for rate limiting, as those are retried with backoff.

File: proto/kv-connect.md (hunk 7 above)
Comment (losfair) [reply proto/kv-connect.md:286]: I think there should be a mechanism to "gracefully" rate limit, like starting to throttle at 90% of the limit and when it reaches 100% wait until the end of the rate limit window

**Pull request merged**